rfcs/0000-top-level-assumptions.md

```diff
@@ -0,0 +1,327 @@
+| `skipForOfIteratorClosing` | When using `for-of` with an iterator, it should always be closed with `.return()` and with `.throw()` in case of an error. This option allows skipping those methods | `loose` in `transform-for-of` | | |
+| `objectRestNoSymbols` | When using rest in object destructuring, don't copy symbol keys. This effectively means using the `objectWithoutPropertiesLoose` helper instead of `objectWithoutProperties` | `loose` in `transform-destructuring` and `proposal-object-rest-spread` _(not documented)_ | [🔗](https://babeljs.io/repl#?browsers=defaults&build=&builtIns=false&spec=false&loose=false&code_lz=G4QwTgBA3hIDQQHTIEYQL4QLwQMYG4g&debug=false&forceAllTransforms=false&shippedProposals=true&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=script&lineWrap=true&presets=env%2Cenv&prettier=false&targets=&version=7.10.1&externalPlugins=) |
+| `setSpreadProperties` | When using object spread, use `Object.assign` to copy the properties instead of cloning their property descriptors with `Object.defineProperty` | `loose`&`useBuiltIns` in `proposal-object-rest-spread` | |
+| `setComputedProperties` | When using computed object properties, use `[[Set]]` semantics (i.e. use an assignment) instead of `[[Define]]` | `loose` in `computed-properties` | [🔗](https://babeljs.io/repl#?browsers=defaults&build=&builtIns=false&spec=false&loose=false&code_lz=G4QwTgBCELwQ3gKAhA2gDwLoC4ICYAaRAXyA&debug=false&forceAllTransforms=false&shippedProposals=true&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=script&lineWrap=true&presets=env%2Cenv&prettier=false&targets=&version=7.10.1&externalPlugins=) |
+| `setClassMethods` | When declaring classes, use `[[Set]]` semantics (i.e. use an assignment) instead of `[[Define]]`. This doesn't preserve the correct enumerability. | `loose` in `transform-classes` | [🔗](https://babeljs.io/repl#?browsers=defaults&build=&builtIns=false&spec=false&loose=true&code_lz=MYGwhgzhAECC0G8BQ1oDMD2GAUBKRAvkgUA&debug=false&forceAllTransforms=false&shippedProposals=true&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=script&lineWrap=true&presets=env%2Cenv&prettier=false&targets=&version=7.10.1&externalPlugins=) |
+| `setPublicClassFields` | When using computed public class fields, use `[[Set]]` semantics (i.e. use an assignment) instead of `[[Define]]` | `loose` in `proposal-class-properties` | [🔗](https://babeljs.io/repl#?browsers=chrome%2070&build=&builtIns=false&spec=false&loose=false&code_lz=MYGwhgzhAECC0G8BQ1oDMCWBTEATaAvNAIwDcSAvkA&debug=false&forceAllTransforms=false&shippedProposals=true&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=script&lineWrap=true&presets=env%2Cenv&prettier=false&targets=&version=7.10.1&externalPlugins=) |
```

Comment:
@existentialism @JLHwung @hzoo Thoughts?

Reply:
I lean to specific assumptions unless we change the interface of assumptions from `{ [string]: boolean }` to `{ [string]: object }`. Assumptions is an advanced feature so I think we should prefer precision over ergonomics, as an advanced user can always add dynamics to assumptions:

```js
const useSetInsteadOfDefine = true;
const assumptions = {
  setPublicClassFields: useSetInsteadOfDefine
}
```

Reply:
I really don't want to use something like `{ [string]: object }` since it makes it harder to reason about options merging.

Whenever we feel like namespacing assumptions (e.g. `{ useSet: { methods: true, fields: true } }`, I would advocated for flattening that object.